File: 2022/02.rs

/**
 * 2022 Day 02: Rock, Paper, Scissor
 */

mod common;

use std::env;
use common::args::check_test_mode;
use common::data::parse_file;

fn main() {
    let year = 2022;
    let day = 2;

    let args: Vec<String> = env::args().collect();
    println!("Advent of Code {year} - Day {day:02}");

    let data: Vec<String> = parse_file(
        String::from(format!("./data/{day:02}")),
        check_test_mode(args),
    );

    // println!("{:?}", data);
    println!("Part 1: {}", part_1(data.clone()));
    // println!("Part 2: {}", part_2(data.clone()));
}

fn part_1(data: Vec<String>) -> i32 {
    let mut res = 0;
    for moves in data.iter() {
        let gestures: Vec<String> = moves.split(' ').map(String::from).collect();
        let elf = &gestures[0];
        let me = &gestures[1];
        let score = compare_chifumi(
            me.to_string(),
            elf.to_string(),
        );
        res += score;
    }
    return res;
}

fn compare_chifumi(a: String, b: String) -> i32 {
    let val_a = get_chifumi_value(a);
    let val_b = get_chifumi_value(b);
    let cmp: i32 = val_a - val_b;
    let score = match cmp {
        0 => 3,
        -1 | -2 => 0,
        1 | 2 => 6,
        _ => 9,
    };
    return score + val_a;
}

fn get_chifumi_value(a: String) -> i32 {
    let keys = vec!["A", "B", "C", "X", "Y", "Z"];
    let values = vec![1, 2, 3, 1, 2, 3];
    let index = keys.iter().position(|&m| m == a);
    return values[index.unwrap()];
}

File: 2022/02.js
/**
 * Share your thoughts
 */

const year = '2022';
const day = '02';
const fs = require('fs');

const parseInput = () => {
    let filename = `./data/${day}`;
    if (process.argv.length > 2) {
        filename += '.test';
    }
    let data = fs.readFileSync(filename, 'utf8');
    return data;
}

const part1 = (data) => {
    console.time('Time');
    let res = 0;

    console.log('Part 1:', res);
    console.timeEnd('Time');
}

const part2 = (data) => {
  console.time('Time');
  let res = 0;

  const pairs = {
    'A X': 3, 'A Y': 6, 'A Z': 0,
    'B X': 0, 'B Y': 3, 'B Z': 6,
    'C X': 6, 'C Y': 0, 'C Z': 3,
  }
  const val = {
    'X': 1, 'Y':2, 'Z': 3,
  }

  data = data.split('\n');
  console.log(data);
  for (const m of data) {
    if (m.length !== 3) {
      continue;
    }
    console.log(pairs[m]);
    res = res + pairs[m] + val[m[2]];
  }
  // 10310: à reporter dans le developpement en RUST !
  console.log('Part 2:', res);
  console.timeEnd('Time');
}

/**
 * Main
 */
console.log( `Advent Of Code ${year}-${day}`);
console.log('==========================');
part1(parseInput());
console.log('--------------------------');
part2(parseInput());
